src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,110 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int INITIAL_CAPACITY = 10;
+    private static final double GROWTH_FACTOR = 1.5;
+    private T[] array;
+    private int size;
+
+    @SuppressWarnings("unchecked")
+    public ArrayList() {
+        array = (T[]) new Object[INITIAL_CAPACITY];
+        size = 0;
+    }
+
+    @SuppressWarnings("unchecked")
+    private void grow() {
+        int newCapacity = (int) (array.length * GROWTH_FACTOR);
+        T[] newArray = (T[]) new Object[newCapacity];
+        System.arraycopy(array, 0, newArray, 0, size);
+        array = newArray;
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new ArrayListIndexOutOfBoundsException("Index " + index
+                    + " out of bounds of size " + size);
+        }
+    }
+
     @Override
     public void add(T value) {
-
+        if (size == array.length) {
+            grow();
+        }
+        array[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index < 0 || index > size) {
```

Comment:
The index validation logic is repeated here. You can extend the 'checkIndex' method to also handle the case when the index is equal to 'size' and use it here to avoid code duplication.